File: graph.py
class Node:
	def __init__(self,data):
		self.data=data
		self.neighbour={}

	def add_neighbour(self,new_neighbour,weight=0):
		self.neighbour[new_neighbour] = weight

	def print_neighbour(self):
		for i in self.neighbour:
			print str(i)+','+str(self.neighbour[i])+' ||| ',

	def get_neighbour(self):
		return self.neighbour

	def get_neighbour_vertex(self):
		l=[]
		for i in self.neighbour:
			l.append(i)
		return l

class Directed_Graph:
	def __init__(self):
		self.vertex={}
		self.total_vertex=0

	def add_vertex(self,data):
		self.total_vertex+=1
		new_vertex = Node(data)
		self.vertex[data] = new_vertex

	def add_directed_edge(self,from_u,to_v,weight=0):
		if from_u not in self.vertex:
			self.add_vertex(from_u)
		if to_v not in self.vertex:
			self.add_vertex(to_v)
		self.vertex[from_u].add_neighbour(to_v,weight)

	def print_directed_graph(self):
		for i in self.vertex:
			print str(i)+" => ",
			self.vertex[i].print_neighbour();
			print ''

	def bfs_util(self,visited,queue,size):
		while size>0:
			u = queue.pop(0)
			size-=1
			print u,
			for i in self.vertex[u].get_neighbour_vertex():
				if visited[i]==0:
					queue.append(i)
					size+=1
					visited[i]=1
		return visited,queue,size
	def initialize_visited(self):
		visited={}
		for i in self.vertex:
			visited[i]=0
		return visited
		
	def bfs(self,source):
		visited = self.initialize_visited()
		queue = [source]
		size = 1
		visited[source]=1
		visited, queue, size = self.bfs_util(visited,queue,size)
		#if graph is disconnected
		for i in visited:
			if visited[i]==0:
				visited[i] = 1
				queue = [i]
				size=1
				visited, queue, size = self.bfs_util(visited,queue,size)
		print ''
	def dfs_util(self,source,visited):
		if visited[source]==0:
			print source,
			visited[source]=1
			for i in self.vertex[source].get_neighbour_vertex():
				visited = self.dfs_util(i,visited)
		return visited

	def dfs(self,source):
		visited=self.initialize_visited()
		visited[source]=1
		print source,
		for i in self.vertex[source].get_neighbour_vertex():
			visited = self.dfs_util(i,visited)
		for i in self.vertex:
			if visited[i]==0:
				visited = self.dfs_util(i,visited)
		print ''

G=Directed_Graph()
G.add_directed_edge(1,2,33)
G.add_directed_edge(1,5,543)
G.add_directed_edge(2,3,3564)
G.add_directed_edge(4,5,434)
G.print_directed_graph()
G.bfs(1)
G.dfs(1)

File: graph.cpp
#include <iostream>
#include <vector>
#include <map>
#include <list>
#include <utility>
#include <deque>
using namespace std;
class Directed_Graph{
public:
	int vertex,edge;
	std::vector< list <pair < int,int> > > graph;
	std::vector<int> size_of_con_comp;
	int size;
	Directed_Graph(int nodes){
		size=0;
		vertex = nodes;
		graph.resize(vertex+1);
	}
	void add_edge(int u,int v,int weight){
		graph[u].push_back(make_pair(v,weight));
	}
	void print_graph(){
		int i;
		for(i=1;i<=vertex;i++){
			cout<<i<<"-> ";
			list <pair <int, int> >:: iterator itr=graph[i].begin();
			while(itr!=graph[i].end()){
				cout<<(itr)->first<<" "<<(itr)->second<<" ";
				itr++;
			}cout<<endl;
		}
	}
	void DFS_util(int source,vector<int >& visited){
		list< pair < int,int > >::iterator itr=graph[source].begin();
		cout<<source<<" ";
		visited[source]=1;
		size++;
		while(itr!=graph[source].end()){
			if(visited[itr->first]==0){
				DFS_util(itr->first,visited);
			}
			itr++;
		}
	}
	void DFS(int source){
		std::vector<int> visited(vertex+1);
		fill(visited.begin(), visited.end(),0);
		DFS_util(source,visited);
		for(int i=1;i<=vertex;i++){
			if(visited[i]==0){
				DFS_util(i,visited);
			}
		}
		cout<<endl;
	}

	void BFS(int source){
		std::vector<int> visited(vertex+1);
		fill(visited.begin(), visited.end(),0);
		std::deque <int> d;
		d.push_back(source);
		visited[source]=1;
		while(!d.empty()){
			int s=d.front();
			//visited[s]=1;
			cout<<s<<" ";
			d.pop_front();
			list <pair <int,int > >::iterator itr = graph[s].begin();
			while(itr!=graph[s].end()){
				if(visited[itr->first]==0){
					d.push_back(itr->first);
					visited[itr->first]=1;
				}
				itr++;
			}
		}
		// if graph is disconnected
		int i;
		for(i=1;i<=vertex;i++){
			if(visited[i]==0){
				d.push_back(i);
				visited[source]=1;
				while(!d.empty()){
					int s=d.front();
					//visited[s]=1;
					cout<<s<<" ";
					d.pop_front();
					list <pair <int,int > >::iterator itr = graph[s].begin();
					while(itr!=graph[s].end()){
						if(visited[itr->first]==0){
							d.push_back(itr->first);
							visited[itr->first]=1;
						}
						itr++;
					}
				}
			}
		}
		cout<<endl;
	}
	void connected_component(){
		int source =1;
		std::vector<int> visited(vertex+1);
		fill(visited.begin(), visited.end(),0);
		size=0;
		DFS_util(source,visited);
		size_of_con_comp.push_back(size);
		size=0;
		for(int i=1;i<=vertex;i++){
			if(visited[i]==0){
				DFS_util(i,visited);
				size_of_con_comp.push_back(size);
				size=0;
			}
		}
		cout<<"total diff conn component "<<size_of_con_comp.size()<<endl;
		for(int i=0;i<size_of_con_comp.size();i++){
			cout<<size_of_con_comp[i]<<" ";
		}cout<<endl;
	}
};

int main(){
	int vertex;
	cin>>vertex;
	Directed_Graph G(vertex);
	G.add_edge(1,2,333);
	G.add_edge(3,2,13);
	G.add_edge(2,3,433);
	G.add_edge(4,5,343);
	G.add_edge(6,7,343);
	G.add_edge(6,8,5435);
	G.print_graph();
	int source=2;
	G.BFS(source);
	G.DFS(source);
	G.connected_component(); // total connected connected and size of each
	return 0;
}